File: 2174-next-greater-numerically-balanced-number/2174-next-greater-numerically-balanced-number.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
using ll = long long;
class Solution {
   public:
    static vector<int> beautiful;
    static vector<int> init() {
        vector<int> list;
        vector<int> freq(10, 0);
        Solution().generate(0, list, freq);
        sort(list.begin(), list.end());
        return list;
    }
    bool isBeautiful(vector<int>& freq) {
        for (int i = 0; i < 10; i++) {
            if (freq[i] > 0 && freq[i] != i) return false;
        }
        return true;
    }
    void generate(ll curr, vector<int>& list, vector<int>& freq) {
        if (curr > 0 && isBeautiful(freq)) {
            list.emplace_back(curr);
        }
        if (curr > 1224444) return;

        for (int i = 1; i <= 6; i++) {
            if (freq[i] < i) {
                freq[i]++;
                generate(curr * 10 + i, list, freq);
                freq[i]--;
            }
        }
    }
    int nextBeautifulNumber(int n) {
        return *upper_bound(beautiful.begin(), beautiful.end(), n);
    }
};
vector<int> Solution::beautiful = Solution::init();

File: 2174-next-greater-numerically-balanced-number/2174-next-greater-numerically-balanced-number.go
func nextBeautifulNumber(n int) int {
	nums := make([]int, 0)
	freq := make([]int, 10)
	generate(0, &nums, freq)
	sort.Ints(nums)

	ub := sort.Search(len(nums), func(i int) bool {
		return nums[i] > n
	})
	return nums[ub]
}

func generate(curr int, nums *[]int, freq []int) {
	if curr > 0 && isBeautiful(freq) {
		*nums = append(*nums, curr)
	}
	if curr > 1224444 {
		return
	}

	for i := 1; i <= 6; i++ {
		if freq[i] < i {
			freq[i]++
			generate(curr*10+i, nums, freq)
			freq[i]--
		}
	}
}

func isBeautiful(freq []int) bool {
	for i := 1; i <= 6; i++ {
		if freq[i] > 0 && freq[i] != i {
			return false
		}
	}
	return true
}
